vcert/connection_tpp_token.py

```diff
@@ -128,6 +149,224 @@ def auth(self):
+            retrive_request['IncludeChain'] = 'false'
+        else:
+            log.error("chain option %s is not valid" % certificate_request.chain_option)
+            raise ClientBadData
+
+        status, data = self._post(URLS.CERTIFICATE_RETRIEVE, data=retrive_request)
+        if status == HTTPStatus.OK:
+            pem64 = data['CertificateData']
+            pem = base64.b64decode(pem64)
+            return parse_pem(pem.decode(), certificate_request.chain_option)
+        elif status == HTTPStatus.ACCEPTED:
+            log.debug(data['Status'])
+            return None
+
+        log.error("Status is not %s. %s" % HTTPStatus.OK, status)
+        raise ServerUnexptedBehavior
+
+    def revoke_cert(self, request):
+        if not (request.id or request.thumbprint):
+            raise ClientBadData
+        d = {
+            "Disable": request.disable
+        }
+        if request.reason:
+            d["Reason"] = request.reason
+        if request.id:
+            d["CertificateDN"] = request.id
+        elif request.thumbprint:
+            d["Thumbprint"] = request.thumbprint
+        else:
+            raise ClientBadData
+        if request.comments:
+            d["Comments"] = request.comments
+        status, data = self._post(URLS.CERTIFICATE_REVOKE, data=d)
+        if status in (HTTPStatus.OK, HTTPStatus.ACCEPTED):
+            return data
+        else:
+            raise ServerUnexptedBehavior
+
+    def renew_cert(self, request, reuse_key=False):
+        if not request.id and not request.thumbprint:
+            log.debug("Request id or thumbprint must be specified for TPP")
+            raise CertificateRenewError
+        if not request.id and request.thumbprint:
+            request.id = self.search_by_thumbprint(request.thumbprint)
+        if reuse_key:
+            log.debug("Trying to renew certificate %s" % request.id)
+            status, data = self._post(URLS.CERTIFICATE_RENEW, data={"CertificateDN": request.id})
+            if not data['Success']:
+                raise CertificateRenewError
+            return
+        cert = self.retrieve_cert(request)
+        cert = x509.load_pem_x509_certificate(cert.cert.encode(), default_backend())
+        for a in cert.subject:
+            if a.oid == x509.NameOID.COMMON_NAME:
+                request.common_name = a.value
+            elif a.oid == x509.NameOID.COUNTRY_NAME:
+                request.country = a.value
+            elif a.oid == x509.NameOID.LOCALITY_NAME:
+                request.locality = a.value
+            elif a.oid == x509.NameOID.STATE_OR_PROVINCE_NAME:
+                request.province = a.value
+            elif a.oid == x509.NameOID.ORGANIZATION_NAME:
+                request.organization = a.value
+            elif a.oid == x509.NameOID.ORGANIZATIONAL_UNIT_NAME:
+                request.organizational_unit = a.value
+        for e in cert.extensions:
+            if e.oid == x509.OID_SUBJECT_ALTERNATIVE_NAME:
```

Comment:
Should URI option be included in here?

Reply:
@LanceGardiner-VEN Eventually, yes, but that's not related to the token auth updates.  The stated requirement for VCert was to support DNS, IP Address, and Email SAN types.  Support for URI and UPN SAN types was recently added to the VCert-Go project based on need from the Venafi partner ecosystem.  If you're interested in contributing the URI SAN support to VCert-Python we'd welcome it 😃